### ISSUE TITLE
Azure Storage Streaming, poison handling when message batch can not be deserialized

### ISSUE THREAD

**Issue by rikbosch** [dotnet/orleans#5394]
Hi,

Using the azure storage streaming provider with the default serializer causes an infinite retry loop of message batches serialized with the previous version.

The problem is most likely in https://github.com/dotnet/orleans/blob/bdaf800326b90c5e0950be6b0a8568add56f3b2c/src/Azure/Orleans.Streaming.AzureStorage/Providers/Streams/AzureQueue/AzureQueueAdapterReceiver.cs#L92

The batchcontainer can not be deserialized and should be considered as `poison`, but instead it keeps trying to read the same batch of events until Expiration Time is reached (which is 7 days by default in azure storage queues).

This situation can be reproduced using the following steps:

1. publish batch of message on a stream
2. shutdown silo (check if message is still in queue)
3. add property to the message
4. start silo with new message format

result is, `AzureQueueAdapterReceiver` throwing at line [L85](https://github.com/dotnet/orleans/blob/bdaf800326b90c5e0950be6b0a8568add56f3b2c/src/Azure/Orleans.Streaming.AzureStorage/Providers/Streams/AzureQueue/AzureQueueAdapterReceiver.cs#L85)

which in turn will throw at [PersistentStreamPullingAgent L435](https://github.com/dotnet/orleans/blob/fd7375eebabd66e3826cd77710b1df13d9c0d3be/src/Orleans.Runtime/Streams/PersistentStream/PersistentStreamPullingAgent.cs#L435)

which will trigger the retry from ` AsyncExecutorWithRetries` 
In the next try, the previous batch is still invisible (due to InvisibilityTimeout, set when retrieving the faulty batch) so it will succeed. 

Eventually our silo was only reading faulty batches and not processing any new messages.

We know that we shouldn't break contracts and that the default Orleans serializer is not version tolerant. We try to avoid making breaking changes, but missed one. It should not have broken down the entire silo I think

**Comment by jason-bragg** [comment]
I agree that a change in data format should not break the processing indefinitely, and we should look into how to resolve this.

In the interim, I suggest introducing a custom IAzureQueueDataAdapter which can catch the deserialization error and mitigate it.

If you wish to modify your event contracts over the lifetime of the product, you 'should' be able to do this with a custom IAzureQueueDataAdapter that uses a version tolerant serializer (bond, protobuf, ..?) to serialize the application events.
